open-apparel-registry/open-apparel-registry
# Index extended fields

## Overview

In order to later enable searching by extended field, adds extended
field values to the facility index when the indexing function is run
and in response to relevant data changes, including splitting and
merging of facilities.

For number_of_workers, creates a string array including any of the
standardized ranges that overlap with any of the submitted range values.

For facility_type, creates a string array of all submitted
facility_types, including ones calculated based on submitted processing
types.

For processing_type, creates a string array of all submitted processing
types, excluding facility_types.

For product_type, creates a `clean()` string array based on submitted
procust_types.

For parent_company_name, creates a string array using the raw data name
for non-matched parent_company, but uses contributor_name for
matched parent_company.

For parent_company_id, creates an integer array using the
contributor_id for matched parent_company.

For native_language_name, creates a unidecoded string array from
submitted native_language_names.

Connects #1587 

## Demo

```
>>> FacilityIndex.objects.filter(name__in=['Alsico', 'ALFA TAPUTFABRIEK NV', 'Aarova vzw']).values('id', 'name', 'number_of_workers', 'facility_type', 'processing_type', 'product_type', 'parent_company_name', 'native_language_name', 'parent_company_id')
<QuerySet [
{'id': 'BE2022046WHE6D5', 'name': 'Aarova vzw', 'number_of_workers': ['Less than 1000', 'More than 10000', '1001-5000'], 'facility_type': ['Final Product Assembly', 'Warehousing / Distribution', 'Printing, Product Dyeing and Laundering'], 'processing_type': ['Extrusion', 'Collecting'], 'product_type': ['hats'], 'parent_company_name': ['ABC Holdingas'], 'native_language_name': ['Hang Zhou Wan Kai Fa Qu Xing Ci Er Lu Yu Bin Hai Er Lu Cha Kou Heng Yuan Gong Ye Yuan Qu '], 'parent_company_id': []}, 
{'id': 'BE20220460EGHZ5', 'name': 'ALFA TAPUTFABRIEK NV', 'number_of_workers': ['1001-5000'], 'facility_type': ['Final Product Assembly', 'Warehousing / Distribution'], 'processing_type': ['Stitching', 'Assembly', 'Cutting', 'Packing', 'Steaming'], 'product_type': ['short sleeved shirts', 'wovens', 'shorts', 'dresses', 'skirts', 'long sleeved shirts', 'jumpsuits', 'trousers'], 'parent_company_name': ['Brand A'], 'native_language_name': [], 'parent_company_id': [4]}, 
{'id': 'BE2022046NPJFZ2', 'name': 'Alsico', 'number_of_workers': ['Less than 1000'], 'facility_type': [], 'processing_type': [], 'product_type': [], 'parent_company_name': [], 'native_language_name': [], 'parent_company_id': []}
]>
```

## Testing Instructions

* Run `./scripts/manage migrate`
* Run `./scripts/server`. Login as `c2@example.com` and submit 
[OAR Extended Fields.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/8073771/OAR.Extended.Fields.csv). Process with `./tools/batch_process {list_id}`
* In `./scripts/manage shell_plus`:
```
FacilityIndex.objects.filter(name__in=['Alsico', 'ALFA TAPUTFABRIEK NV', 'Aarova vzw']).values('id', 'name', 'number_of_workers', 'facility_type', 'processing_type', 'product_type', 'parent_company_name', 'native_language_name', 'parent_company_id')
```
* Compare the values of the indexed items to the values submitted in the CSV to confirm they match, all relevant values are included, and values aren't repeated. 
* Authenticated at http://localhost:8081/api/docs/#!/facilities/facilities_create and post:
```
{
"name": "ALFA TAPUTFABRIEK NV",
"address": "Lindestraat 96, B-8790 Waregem , Belgium",
"country": "Belgium",
"parent_company": "Brand A",
"processing_type": ["Cutting", "Stitching", "Steaming"]
}
```
* Run the same facility index query in `shell_plus` as before. Confirm that the new values from the API submission have been added to ALFA TAPUTFABRIEK NV.
* Claim 'Aarova vzw' as c2@example.com
* In another browser session, login as c1@example.com and approve the claim. 
* Update the claim details in 'My Facilities' as c2@example.com. Submit "杭州湾开发区兴慈二路与滨海二路叉口恒元工业园区" as the native language name (to test that a long unidecoded name works) in addition to filling the other claim fields that we process for extended fields.
* Run the same facility index query in `shell_plus` as before. Confirm that the new values from the claim have been added to Aarova vzw.
* Get the ID for ALFA TAPUTFABRIEK NV. Login as c1@example.com and navigate to `Adjust facility matches`. Enter the id and split ALFA TAPUTFABRIEK NV.
* Run the same facility index query in `shell_plus` as before. Confirm that there are now two facilities named ALFA TAPUTFABRIEK NV. One should contain the extended field values from the CSV item and the other should contain the extended field values from the API submission. 
* Navigate to 'Merge two facilities' and merge the two facilities named ALFA TAPUTFABRIEK NV. 
* Run the same facility index query in `shell_plus` as before. Confirm that there is now only one facility named ALFA TAPUTFABRIEK NV and that it contains both sets of extended field values. 

## Checklist
